### PR TITLE
Directly access ICacheService through hubclient+activation args to entirely eliminate UI dependency.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/VisualStudio/Core/Def/Storage/AbstractCloudCachePersistentStorageService.cs [hunk 1]
```diff
@@ -40,13 +39,17 @@ protected sealed override bool ShouldDeleteDatabase(Exception exception)
         protected sealed override async ValueTask<IChecksummedPersistentStorage?> TryOpenDatabaseAsync(
             SolutionKey solutionKey, string workingFolderPath, string databaseFilePath, CancellationToken cancellationToken)
         {
-            var cacheService = await this.CreateCacheServiceAsync(cancellationToken).ConfigureAwait(false);
+            var solutionFolder = IOUtilities.PerformIO(() => Path.GetDirectoryName(solutionKey.FilePath));
+
+            if (solutionFolder == null || solutionFolder == "")
+                return null;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Storage/AbstractCloudCachePersistentStorageService.cs:45]: using `string.IsNullOrEmpty` does not seem to prove to the compiler that the string is non-null after this.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Storage/AbstractCloudCachePersistentStorageService.cs:45]: (this file is shared, and it looks like in some configs this isn't nrt annotated).

File: src/Tools/IdeCoreBenchmarks/CloudCache/IdeCoreBenchmarksCloudCacheServiceProvider.cs [hunk 2]
```diff
@@ -25,17 +25,7 @@ public IdeCoreBenchmarksCloudCacheServiceProvider()
...
-                    if (cs is IAsyncDisposable asyncDisposable)
-                    {
-                        asyncDisposable.DisposeAsync().AsTask().Wait();
-                    }
-                    else if (cs is IDisposable disposable)
-                    {
-                        disposable.Dispose();
-                    }
-                });
+                configuration, @"C:\github\roslyn");
```
Comment (CyrusNajmabadi) [inline src/Tools/IdeCoreBenchmarks/CloudCache/IdeCoreBenchmarksCloudCacheServiceProvider.cs:28]: have gotten confirmation that we only need to support .Dispose for ICacheService proxies, so this all became much simpler.

File: src/VisualStudio/CSharp/Test/PersistentStorage/CloudCachePersistentStorageTests.cs [hunk 3]
```diff
@@ -18,22 +14,7 @@ public class CloudCachePersistentStorageTests : AbstractPersistentStorageTests
...
-            var threadingContext = exportProvider.GetExports<IThreadingContext>().Single().Value;
-            return new MockCloudCachePersistentStorageService(
-                configuration,
-                relativePathBase,
-                cs =>
-                {
-                    if (cs is IAsyncDisposable asyncDisposable)
-                    {
-                        threadingContext.JoinableTaskFactory.Run(
-                            () => asyncDisposable.DisposeAsync().AsTask());
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/CSharp/Test/PersistentStorage/CloudCachePersistentStorageTests.cs:30]: no need to support sync over async, so this all became trivial.

File: src/VisualStudio/Core/Def/Storage/HubClientCloudCachePersistentStorageService.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            // Directly use HubClient.  This allows us to avoid everything related to ServiceBroker, which keeps us from
+            // ever hitting the UI thread.
+            using var hubClient = new HubClient();
+
+#pragma warning disable ISB001 // Dispose of proxies
+            // cache service will be disposed inside WrappedCacheService.Dispose
+            var cacheService = await hubClient.GetProxyAsync<ICacheService>(
+                VisualStudioServices.VS2019_10.CacheService,
+                new ServiceActivationOptions { ActivationArguments = new Dictionary<string, string> { { "foo", solutionFolder } } },
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Storage/HubClientCloudCachePersistentStorageService.cs:33]: foo will be replaced with the actual key name once we know it.

Comment (AArnott) [reply src/VisualStudio/Core/Def/Storage/HubClientCloudCachePersistentStorageService.cs:33]: This will be `RelativePathBase`

Comment (AArnott) [reply src/VisualStudio/Core/Def/Storage/HubClientCloudCachePersistentStorageService.cs:33]: And `ImmutableDictionary<string, string>.Empty.Add("RelativePathBase", solutionFolder)` may result in fewer/smaller allocations.

**Comment by CyrusNajmabadi** [comment]
Closing this out.  We can't use hubclient due to authorization issues.

**Pull request closed without merging**